flink-connectors/flink-connector-hbase/src/main/java/org/apache/flink/connector/hbase/util/HBaseSerde.java

```diff
@@ -0,0 +1,371 @@
+					put.addColumn(familyKey, qualifier, value);
+				}
+			}
+		}
+		return put;
+	}
+
+	/**
+	 * Returns an instance of Delete that remove record from HBase table.
+	 *
+	 * @return The appropriate instance of Delete for this use case.
+	 */
+	public @Nullable Delete createDeleteMutation(RowData row) {
+		byte[] rowkey = keyEncoder.encode(row, rowkeyIndex);
+		if (rowkey.length == 0) {
+			// drop dirty records, rowkey shouldn't be zero length
+			return null;
+		}
+		// delete
+		Delete delete = new Delete(rowkey);
+		for (int i = 0; i < fieldLength; i++) {
+			if (i != rowkeyIndex) {
+				int f = i > rowkeyIndex ? i - 1 : i;
+				// get family key
+				byte[] familyKey = families[f];
+				for (int q = 0; q < this.qualifiers[f].length; q++) {
+					// get quantifier key
+					byte[] qualifier = qualifiers[f][q];
+					delete.addColumn(familyKey, qualifier);
+				}
+			}
+		}
+		return delete;
+	}
+
+	/**
+	 * Returns an instance of Scan that retrieves the required subset of records from the HBase table.
+	 *
+	 * @return The appropriate instance of Scan for this use case.
+	 */
+	public Scan createScan() {
+		Scan scan = new Scan();
+		for (int f = 0; f < families.length; f++) {
+			byte[] family = families[f];
+			for (int q = 0; q < qualifiers[f].length; q++) {
+				byte[] quantifier = qualifiers[f][q];
+				scan.addColumn(family, quantifier);
+			}
+		}
+		return scan;
+	}
+
+	/**
+	 * Converts HBase {@link Result} into {@link RowData}.
+	 */
+	public RowData convertToRow(Result result) {
+		Object rowkey = keyDecoder.decode(result.getRow());
+		for (int i = 0; i < fieldLength; i++) {
+			if (rowkeyIndex == i) {
+				reusedRow.setField(rowkeyIndex, rowkey);
+			} else {
+				int f = (rowkeyIndex != -1 && i > rowkeyIndex) ? i - 1 : i;
+				// get family key
+				byte[] familyKey = families[f];
+				GenericRowData familyRow = reusedFamilyRows[f];
+				for (int q = 0; q < this.qualifiers[f].length; q++) {
+					// get quantifier key
+					byte[] qualifier = qualifiers[f][q];
+					// read value
+					byte[] value = result.getValue(familyKey, qualifier);
+					familyRow.setField(q, qualifierDecoders[f][q].decode(value));
+				}
+				reusedRow.setField(i, familyRow);
+			}
+		}
+		return reusedRow;
+	}
+
+	// ------------------------------------------------------------------------------------
+	// HBase Runtime Encoders
+	// ------------------------------------------------------------------------------------
+
+	/**
+	 * Runtime encoder that encodes a specified field in {@link RowData} into byte[].
+	 */
+	@FunctionalInterface
+	private interface FieldEncoder extends Serializable {
+		byte[] encode(RowData row, int pos);
+	}
+
+	private static FieldEncoder createNullableFieldEncoder(LogicalType fieldType) {
+		final FieldEncoder encoder = createFieldEncoder(fieldType);
+		if (fieldType.isNullable()) {
+			return (row, pos) -> {
+				if (row.isNullAt(pos)) {
+					return EMPTY_BYTES;
+				} else {
```

Comment:
For string, use a 'N/A' seems better which has no conflicts with empty string `""`.

Reply:
I will use "null" which is also used in elasticsearch `null-key-literal`.